### PR TITLE
Add rotation to EllipseCurve and Path.

### PR DESCRIPTION
**Pull request by instacat**
Adds an optional rotation parameter to THREE.EllipseCurve, THREE.Path.ellipse and THREE.Path.absellipse.

### CODE REVIEW

**Comment by WestLangley** [comment]
This seems reasonable. Thanks.

@neko1235 Would you mind looking into the following issue:

> When going clockwise it's best to set the start angle to (Math.PI * 2) and then work towards lower numbers.

This has caused a lot of confusion. Would it be correct to say this, instead?

If `clockwise` is false (the default), set `0 < startAngle < endAngle < 2 * Math.PI`.

If `clockwise` is true, set `2 * Math.PI > startAngle > endAngle > 0`.

If that is not correct, we need to modify the code so the behavior is understandable.

File: docs/api/extras/core/Path.html [hunk 1]
```diff
@@ -1,123 +1,125 @@
-<!DOCTYPE html>
-<html lang="en">
-	<head>
+<!DOCTYPE html>
+<html lang="en">
```
Comment (WestLangley) [inline docs/api/extras/core/Path.html]: Why is almost every line changed in this file?

Comment (neko1235) [inline docs/api/extras/core/Path.html]: I have no idea... I changed only a few lines.
Would you like me to redo this change?

Comment (WestLangley) [inline docs/api/extras/core/Path.html]: It is worth a try.

Comment (dubejf) [inline docs/api/extras/core/Path.html]: It's dos/unix end of line inconsistency. Same issue here https://github.com/mrdoob/three.js/pull/6859#issuecomment-122063407

**Comment by neko1235** [comment]
@WestLangley Please take another look.

**Comment by neko1235** [comment]
@WestLangley 
> When going clockwise it's best to set the start angle to (Math.PI * 2) and then work towards lower numbers.

My understanding is that `startAngle > endAngle` changes drawing direction from that defined by `clockwise` to the opposite.
> If that is not correct, we need to modify the code so the behavior is understandable.

I'm not sure what are the rules here... Changing the behavior will make `EllipseCurve` not backwards compatible.

**Comment by WestLangley** [comment]
This looks fine. `Path.js` needs some clean-up later.

> I'm not sure what are the rules here...

We need to fix it. We can deal with it in another PR. If you would like to spend some time experimenting, that would be great.

**Comment by neko1235** [comment]
Thank you @WestLangley and @dubejf.
Do you want to open a bug where we could figure out what should be the correct behavior of ‘EllipseCurve‘ and/or its doc?

**Comment by WestLangley** [comment]
@neko1235 Please do.

**Comment by neko1235** [comment]
@WestLangley Sure. Thanks.

File: src/extras/core/Path.js [hunk 2]
```diff
@@ -386,13 +386,22 @@ THREE.Path.prototype.getPoints = function( divisions, closedPath ) {
...
+				aClockwise = !! args[ 6 ],
+				aRotation = args[ 7 ] || 0;
 
 
 			var deltaAngle = aEndAngle - aStartAngle;
 			var angle;
 			var tdivisions = divisions * 2;
 
+			var cos, sin;
+			if ( aRotation ) {
```
Comment (mrdoob) [inline src/extras/core/Path.js]: At first I though `aRotation` was a boolean.

I think this check should be `if ( aRotation !== 0 ) {` instead.

File: src/extras/core/Path.js [hunk 3]
```diff
@@ -408,6 +417,14 @@ THREE.Path.prototype.getPoints = function( divisions, closedPath ) {
 				tx = aX + xRadius * Math.cos( angle );
 				ty = aY + yRadius * Math.sin( angle );
 
+				if ( aRotation ) {
```
Comment (mrdoob) [inline src/extras/core/Path.js]: Same here... `if ( aRotation !== 0 ) {`.

File: src/extras/curves/EllipseCurve.js [hunk 4]
```diff
@@ -39,11 +41,20 @@ THREE.EllipseCurve.prototype.getPoint = function ( t ) {
 
 	}
 	
-	var vector = new THREE.Vector2();
+	var x = this.aX + this.xRadius * Math.cos( angle );
+	var y = this.aY + this.yRadius * Math.sin( angle );
+
+	if ( this.aRotation ) {
```
Comment (mrdoob) [inline src/extras/curves/EllipseCurve.js]: Same here... `if ( this.aRotation !== 0 ) {`.

**Comment by neko1235** [comment]
@mrdoob Thank you. Please take another look.

**Comment by mrdoob** [comment]
Looks good!

**Comment by mrdoob** [comment]
Thanks!

**Comment by neko1235** [comment]
@mrdoob Thank you!

**Comment by WestLangley** [comment]
Unfortunately, this PR has seriously broken the examples `webgl_geometry_shapes.html` and `canvas_geometry_shapes.html`.

@neko1235 Can you please have a look?

**Comment by neko1235** [comment]
:sob:

File: src/extras/core/Path.js [hunk 5]
```diff
@@ -408,6 +417,14 @@ THREE.Path.prototype.getPoints = function( divisions, closedPath ) {
 				tx = aX + xRadius * Math.cos( angle );
 				ty = aY + yRadius * Math.sin( angle );
 
+				if ( aRotation !== 0 ) {
+
+					// Rotate the point about the center of the ellipse.
+					tx = ( tx - aX ) * cos - ( ty - aY ) * sin + aX;
+					ty = ( tx - aX ) * sin + ( ty - aY ) * cos + aY;
```
Comment (mrdoob) [inline src/extras/core/Path.js]: I suspect this is wrong, actually. `tx` and `ty` are being changed while being used.

Comment (WestLangley) [inline src/extras/core/Path.js]: Hmmm... I do not see a problem with that...

Comment (mrdoob) [inline src/extras/core/Path.js]: Shouldn't it be something like this instead?

```javascript
// Rotate the point about the center of the ellipse.
var x = tx, y = ty;
tx = ( x - aX ) * cos - ( y - aY ) * sin + aX;
ty = ( x - aX ) * sin + ( y - aY ) * cos + aY;
```

Comment (neko1235) [inline src/extras/core/Path.js]: Huh... yes this is wrong...

File: src/extras/core/Path.js [hunk 6]
```diff
@@ -153,24 +153,24 @@ THREE.Path.prototype.arc = function ( aX, aY, aRadius,
...
+		aStartAngle, aEndAngle, aClockwise, aRotation );
 
  };
 
 
 THREE.Path.prototype.absellipse = function ( aX, aY, xRadius, yRadius,
-									  aStartAngle, aEndAngle, aClockwise ) {
+									  aStartAngle, aEndAngle, aClockwise, aRotation ) {
 
 	var args = Array.prototype.slice.call( arguments );
```
Comment (neko1235) [inline src/extras/core/Path.js]: This doesn't work well when there are optional arguments.
Can I replace this with something like this?
```javascript
var args = [ aX, aY, xRadius, yRadius, aStartAngle, aEndAngle, aClockwise, aRotation || 0 ];
```
What do you think?

Comment (WestLangley) [inline src/extras/core/Path.js]: Yes, you found the issue. I am not sure of the best solution. `Path.js` is fragile, and is due for a re-design. Please post what you feel is best in a new PR and we can have a look. It needs to be backwards-compatible. Maybe @mrdoob can provide some guidance.

**Pull request merged**